network/src/network.rs

```diff
@@ -255,54 +247,57 @@ impl NetworkState {
         protocol_id: ProtocolId,
         protocol_version: ProtocolVersion,
     ) -> Result<RegisterResult, Error> {
-        let mut peers_registry = self.peers_registry.write();
         let register_result = if session_type.is_outbound() {
-            peers_registry.try_outbound_peer(
+            self.peers_registry.try_outbound_peer(
                 peer_id.clone(),
                 connected_addr,
                 session_id,
                 session_type,
             )
         } else {
-            peers_registry.accept_inbound_peer(
+            self.peers_registry.accept_inbound_peer(
                 peer_id.clone(),
                 connected_addr,
                 session_id,
                 session_type,
             )
         }?;
         // add session to peer
-        match peers_registry.get_mut(&peer_id) {
-            Some(peer) => match peer.protocol_version(protocol_id) {
-                Some(_) => return Err(ProtocolError::Duplicate(protocol_id).into()),
+        self.peers_registry
```

Comment:
What if other thread update `peers_registry` before here and after `register_result`?